Cluster Template ref guide (#158)
AxonIQ/reference-guide: axon-server/administration/cluster-template.md

> #Cluster template

Reviewer: Don't forget to add this page to the `SUMMARY.md`, otherwise it will not show up in the index for our readers.

Reviewer: Additionally, it should be added to an Axon Server specific section. Thinking `installation` would be the most likely, thus warranting another change in the `installation/README.md` file.

Author: I added it to configuration folder, next to the CLI, REST configuration

> With the Cluster Template feature we enabled users to maintain cluster configuration as a code.

Reviewer: ```suggestion
With the Cluster Template feature we enable users to maintain cluster configuration as code.
```

> Cluster template is defined as a YAML file which describes cluster configuration. It is possible to pre-define replication groups, contexts, metadata, applications (with tokens), users and their roles and that configuration can be shared across teams.

Reviewer: ```suggestion
The cluster template is defined as a YAML file, describing a cluster's configuration. It is possible to predefine replication groups, contexts, metadata, applications (with tokens), users and their roles, so that the configuration can be shared across teams.
```

{% hint style="warn" %}

> Cluster template runs exactly once, on the first clean Axon Server start-up, if there is no previous cluster configuration defined. Therefore, cluster template WILL NOT override any existing configuration, its purpose is to be used during active development and to share cluster configuration across the developing team.

Reviewer: ```suggestion
The cluster template runs exactly once, on the first clean Axon Server start-up, if there is no previous cluster configuration defined. Therefore, the cluster template **will not** override any existing configuration. Its purpose is to be used during active development, to be able to share the configuration across development teams.
```

{% endhint %}

{% hint style="warn" %}

> As clustering is an Axon Server EE feature, license needs to be provided before using this feature.

Reviewer: ```suggestion
As clustering is an Axon Server EE feature, a license needs to be provided to be able to use this feature.
```

{% endhint %}

## Usage

> To use Cluster template all you need to do is define a valid cluster template yml file. If this file is present on a clean Axon Server startup, it will automatically be picked up, and the cluster will be configured accordingly.

Reviewer: ```suggestion
To use the cluster template feature, all you need to do is define a valid cluster template yml file. If this file is present on a fresh Axon Server startup, it will automatically be picked up and the cluster will be configured accordingly.
```

{% hint style="warn" %}

> Each cluster node needs to have the cluster template yml file copy. Each node will read this file, find its own configuration and configure itself.

Reviewer: ```suggestion
Each cluster node needs to have a cluster template yml file copy. Each node will read this file, find its own configuration and configure itself.
```

{% endhint %}

> Default path from which Axon Server reads configuration is `./cluster-template.yml`

Reviewer: ```suggestion
The default path from which Axon Server reads this configuration file is `./cluster-template.yml`.
```

You can override this path anytime by setting Axon Server property: 
`axoniq.axonserver.clustertemplate.path:/mypath/cluster-template.yml`

### Configuration

> Below you can find an example of basic cluster setup: _admin & default context in separate replication nodes, replicated across all nodes that are marked as primary.

Reviewer: ```suggestion
Below you can find an example of a basic cluster setup: the _admin and default contexts are in separate replication nodes, replicated across all nodes that are marked as primary.
```

```yaml
axoniq:
  axonserver:
    cluster-template:
      first: axonserver-enterprise-1
      replicationsGroups:
      - name: _admin
        roles:
        - node: axonserver-enterprise-1
          role: PRIMARY
        - node: axonserver-enterprise-2
          role: PRIMARY
        - node: axonserver-enterprise-3
          role: PRIMARY
        contexts:
        - name: _admin
      - name: default
        roles:
        - node: axonserver-enterprise-2
          role: PRIMARY
        - node: axonserver-enterprise-3
          role: PRIMARY
        - node: axonserver-enterprise-1
          role: PRIMARY
        contexts:
        - name: default
      applications: []
      users: []
```

![Cluster overview after default configuration is applied](/.gitbook/assets/cluster-template-default-configuration.png)
_Cluster overview after default configuration is applied_

### Export/Generator

> In order to avoid mistakes while writing with cluster configuration we implemented export button that will generate cluster template file based on current cluster configuration. 

Reviewer: ```suggestion
In order to avoid mistakes while writing a cluster configuration file, we have implemented an export button that will generate a cluster template file based on current setup. 
```

![Cluster Template export button location](/.gitbook/assets/cluster-template-export-button.png)
_Location of export button at Settings page_

> **Recommended way for creating advanced cluster setup**

Reviewer: ```suggestion
**Recommended way for creating an advanced cluster setup**
```

> * Start a fresh Axon Server (use basic cluster template setup mentioned above)
> * Configure a cluster via UI by creating users, applications, replication groups & contexts
> * Use export button located at Settings tab > Configuration panel  to download current cluster configuration
> * Replace basic cluster template with newly exported cluster template configuration

Reviewer: ```suggestion
* Start a fresh Axon Server setup (use basic cluster template setup mentioned above).
* Configure a cluster via the UI, by creating users, applications, replication groups and contexts.
* Use the export button located at "Settings -> Configuration" panel  to download the current cluster configuration.
* Replace the basic cluster template with the newly exported cluster template configuration.
```

{% hint style="warn" %}

> Use export button from main - leader node.

Reviewer: ```suggestion
Use the export button from the main / leader node.
```

Reviewer: Would you mind elaborating why this is recommended? I have a hunch, but would be valuable for the readers too.

Author: One part is mentioned above:

_In order to avoid mistakes while writing a cluster configuration file, we have implemented an export button that will generate a cluster template file based on current setup._ 

But the real reason is:

Ideally, we would have cluster configuration generator on our website, that generates cluster-template.yml. Since we don't, right now, user needs somehow to create configuration before exporting it as cluster-template.yml to be used by the rest of the team, therefor are these recommended steps. It's easy to explain to you but I don't know how to write this down to ref guide... Any hints?

Alternative: skip this section and ask Dusan to create FE - cluster template generator. Would be a great feature for website

Reviewer: Aah, what I mean is _why_ the export method should be called from a leader/main node specifically. It makes sense, but might be helpful for our readers.
